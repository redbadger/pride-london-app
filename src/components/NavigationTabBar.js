// @flow
import React from "react";
import { Animated, StyleSheet, View } from "react-native";
import type { TabScene, _TabBarBottomProps } from "react-navigation";
import SafeAreaView from "react-native-safe-area-view";
import {
  tabBarBgColor,
  tabBarLabelColor,
  tabBarActiveLineColor,
  tabBarActiveLabelColor,
  tabBarBorderColor
} from "../constants/colors";
import Text from "./Text";
import Touchable from "./Touchable";

class NavigationTabBar extends React.PureComponent<_TabBarBottomProps> {
  constructor() {
    super();
    this.tabBarWidth = new Animated.Value(1);
    this.tabLayouts = {};
    this.activeTabX = new Animated.Value(0);
    this.activeTabWidth = new Animated.Value(0);
  }

  componentDidUpdate() {
    this.updateActiveTabLine();
  }

  tabBarLayout: any;
  tabBarWidth: Animated.Value;
  tabLayouts: { [number]: any };
  activeTabX: Animated.Value;
  activeTabWidth: Animated.Value;

  handleTabPress = (key: string) => {
    this.props.jumpTo(key);
  };

  updateActiveTabLine = () => {
    const { navigation } = this.props;
    const currentIndex = navigation.state.index;
    const layout = this.tabLayouts[currentIndex];
    if (!layout || !this.tabBarLayout) {
      return;
    }

    Animated.timing(this.tabBarWidth, {
      toValue: this.tabBarLayout.width,
      duration: 150,
      useNativeDriver: true
    }).start();
    Animated.timing(this.activeTabX, {
      toValue: layout.x,
      duration: 150,
      useNativeDriver: true
    }).start();
    Animated.timing(this.activeTabWidth, {
      toValue: layout.width,
      duration: 150,
      useNativeDriver: true
    }).start();
  };

  renderLabel = (scene: TabScene) => {
    const { focused } = scene;
    const color = focused ? tabBarActiveLabelColor : tabBarLabelColor;
    const label = this.props.getLabelText(scene);

    return (
      <Text type="tabBarItem" numberOfLines={1} style={{ color }}>
        {label}
      </Text>
    );
  };

  renderIcon = (scene: TabScene) => {
    const { route, index, focused } = scene;
    const opacity = focused ? 1 : 0;
    const opacitInv = focused ? 0 : 1;

    // We render the icon twice at the same position on top of each other:
    // active and inactive one, so we can fade between them.
    return (
      <View style={styles.iconContainer}>
        <View style={[styles.icon, { opacity }]}>
          {this.props.renderIcon({
            route,
            index,
            focused: true
          })}
        </View>
        <View style={[styles.icon, { opacity: opacitInv }]}>
          {this.props.renderIcon({
            route,
            index,
            focused: false
          })}
        </View>
      </View>
    );
  };

  render() {
    const { navigation, getTabTestID } = this.props;
    const { routes } = navigation.state;

    const activeTabLineScale = Animated.divide(
      this.activeTabWidth,
      this.tabBarWidth
    );
    const activeTabLineTranslate = Animated.add(
      this.activeTabX,
      Animated.add(
        Animated.divide(this.tabBarWidth, -2),
        Animated.divide(this.activeTabWidth, 2)
      )
    );

    return (
      <SafeAreaView
        style={styles.tabBarWrapper}
        forceInset={{ bottom: "always", top: "never" }}
      >
        <View
          testID="tab-bar"
          onLayout={event => {
            this.tabBarLayout = event.nativeEvent.layout;
            this.updateActiveTabLine();
          }}
          style={styles.tabBar}
        >
          {routes.map((route, index) => {
            const focused = index === navigation.state.index;
            const scene = { route, index, focused };
            return (
              <Touchable
                key={route.key}
                testID={getTabTestID(route.routeName)}
                onLayout={event => {
                  this.tabLayouts[index] = event.nativeEvent.layout;
                  this.updateActiveTabLine();
                }}
                onPress={() => this.handleTabPress(route.key)}
                style={styles.tab}
              >
                {this.renderIcon(scene)}
                {this.renderLabel(scene)}
              </Touchable>
            );
          })}
          <Animated.View
            style={[
              styles.activeTabLine,
              {
                transform: [
                  { translateX: activeTabLineTranslate },
                  { scaleX: activeTabLineScale }
                ]
              }
            ]}
          />
        </View>
      </SafeAreaView>
    );
  }
}

const styles = StyleSheet.create({
  tabBarWrapper: {
    backgroundColor: tabBarBgColor,
    borderTopWidth: StyleSheet.hairlineWidth,
    borderTopColor: tabBarBorderColor,
    paddingHorizontal: 8
  },
  activeTabLine: {
    backgroundColor: tabBarActiveLineColor,
    height: 3,
    position: "absolute",
    left: 0,
    top: 0,
    width: "100%"
  },
  tabBar: {
    flexDirection: "row",
    height: 52,
    width: "100%",
    maxWidth: 440,
    alignSelf: "center"
  },
  tab: {
    alignItems: "center",
    flexDirection: "column",
    minWidth: 48,
    flexGrow: 1,
    marginHorizontal: 4,
    paddingTop: 4,
    paddingBottom: 2
  },
  iconContainer: {
    flex: 1,
    width: "100%"
  },
  icon: {
    position: "absolute",
    alignItems: "center",
    alignSelf: "center",
    top: 0
  }
});

export default NavigationTabBar;
