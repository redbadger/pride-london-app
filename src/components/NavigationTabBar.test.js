// @flow
import React from "react";
import { Animated, Text } from "react-native";
import { shallow } from "enzyme";
import NavigationTabBar from "./NavigationTabBar";

let timingSpy;
const navigationDispatch = jest.fn();

const render = props =>
  shallow(
    <NavigationTabBar
      position={new Animated.Value(0)}
      navigation={{
        dispatch: navigationDispatch,
        state: {
          index: 0,
          routes: [
            { key: "1", index: 1, routeName: "tab-0" },
            { key: "2", routeName: "tab-1" }
          ]
        }
      }}
      jumpTo={() => {}}
      getTabTestID={routeName => routeName}
      renderIcon={scene => <Text>{scene.index}</Text>}
      getLabelText={scene => `Tab ${scene.index}`}
      {...props}
    />
  );

beforeEach(() => {
  timingSpy = jest.spyOn(Animated, "timing").mockImplementation(() => ({
    start: jest.fn()
  }));
});

it("renders correctly", () => {
  const output = render();
  expect(output).toMatchSnapshot();
});

it("stores measurements of tab bar onLayout", () => {
  const output = render();
  const layout = { width: 100, height: 200 };
  output
    .find({ testID: "tab-bar" })
    .props()
    .onLayout({
      nativeEvent: { layout }
    });
  expect(output.instance().tabBarLayout).toBe(layout);
});

it("stores measurements of tabs onLayout", () => {
  const output = render();
  const layout = { width: 100, height: 200 };
  output
    .find({ testID: "tab-0" })
    .props()
    .onLayout({
      nativeEvent: { layout }
    });
  expect(output.instance().tabLayouts[0]).toBe(layout);
});

it("jumps to tab onPress", () => {
  const jumpTo = jest.fn();
  const output = render({ jumpTo });
  output
    .find({ testID: "tab-1" })
    .props()
    .onPress();
  expect(jumpTo).toHaveBeenCalledWith("2");
});

it("animates active tab line on focus change", () => {
  const output = render();
  const instance = output.instance();
  instance.tabBarLayout = { x: 50, width: 400 };
  instance.tabLayouts = {
    "0": { x: 0, width: 48 },
    "1": { x: 60, width: 48 }
  };
  instance.componentDidUpdate();

  expect(timingSpy).toHaveBeenCalledTimes(3);
  expect(timingSpy).toHaveBeenCalledWith(instance.tabBarWidth, {
    toValue: instance.tabBarLayout.width,
    duration: 150,
    useNativeDriver: true
  });
});

afterEach(() => {
  timingSpy.mockRestore();
  navigationDispatch.mockClear();
});
