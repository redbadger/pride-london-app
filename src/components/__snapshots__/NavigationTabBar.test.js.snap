// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<withOrientation
  forceInset={
    Object {
      "bottom": "always",
      "top": "never",
    }
  }
  style={
    Object {
      "backgroundColor": "rgb(243, 243, 243)",
      "borderTopColor": "rgba(0, 0, 0, 0.25)",
      "borderTopWidth": 1,
      "paddingHorizontal": 8,
    }
  }
>
  <Component
    onLayout={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "flexDirection": "row",
        "height": 52,
        "maxWidth": 440,
        "width": "100%",
      }
    }
    testID="tab-bar"
  >
    <Touchable
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      delayPressIn={50}
      key="1"
      onLayout={[Function]}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "flexGrow": 1,
          "marginHorizontal": 4,
          "minWidth": 48,
          "paddingBottom": 2,
          "paddingTop": 4,
        }
      }
      testID="tab-0"
    >
      <Component
        style={
          Object {
            "flex": 1,
            "width": "100%",
          }
        }
      >
        <Component
          style={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "position": "absolute",
                "top": 0,
              },
              Object {
                "opacity": 1,
              },
            ]
          }
        >
          <Component>
            0
          </Component>
        </Component>
        <Component
          style={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "position": "absolute",
                "top": 0,
              },
              Object {
                "opacity": 0,
              },
            ]
          }
        >
          <Component>
            0
          </Component>
        </Component>
      </Component>
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        numberOfLines={1}
        style={
          Object {
            "color": "rgb(45, 47, 127)",
          }
        }
        type="tabBarItem"
      >
        Tab 0
      </Text>
    </Touchable>
    <Touchable
      accessibilityComponentType="button"
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      delayPressIn={50}
      key="2"
      onLayout={[Function]}
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "flexGrow": 1,
          "marginHorizontal": 4,
          "minWidth": 48,
          "paddingBottom": 2,
          "paddingTop": 4,
        }
      }
      testID="tab-1"
    >
      <Component
        style={
          Object {
            "flex": 1,
            "width": "100%",
          }
        }
      >
        <Component
          style={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "position": "absolute",
                "top": 0,
              },
              Object {
                "opacity": 0,
              },
            ]
          }
        >
          <Component>
            1
          </Component>
        </Component>
        <Component
          style={
            Array [
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "position": "absolute",
                "top": 0,
              },
              Object {
                "opacity": 1,
              },
            ]
          }
        >
          <Component>
            1
          </Component>
        </Component>
      </Component>
      <Text
        color="blackColor"
        markdown={false}
        markdownStyle={Object {}}
        numberOfLines={1}
        style={
          Object {
            "color": "rgb(88, 88, 88)",
          }
        }
        type="tabBarItem"
      >
        Tab 1
      </Text>
    </Touchable>
    <AnimatedComponent
      style={
        Array [
          Object {
            "backgroundColor": "rgb(44, 218, 157)",
            "height": 3,
            "left": 0,
            "position": "absolute",
            "top": 0,
            "width": "100%",
          },
          Object {
            "transform": Array [
              Object {
                "translateX": -0.5,
              },
              Object {
                "scaleX": 0,
              },
            ],
          },
        ]
      }
    />
  </Component>
</withOrientation>
`;
